tcl/pdtk_canvas.tcl

```diff
@@ -503,3 +503,351 @@ proc ::pdtk_canvas::cords_to_foreground {mytoplevel {state 1}} {
+    set out(x1) [lindex $args 1]
+    set out(y1) [lindex $args 2]
+    set out(x2) [lindex $args 3]
+    set out(y2) [lindex $args 4]
+}
+
+proc parse_poly_types_args {args argc type outPtr} {
+    upvar 1 $outPtr out
+    set defaultCapstyle "-capstyle projecting"
+    if { $type in {floatatom symbolatom listbox msg} } {
+        check_argc_exact 8 $argc
+        set out(corner) [lindex $args 5]
+        set out(pattern) ""
+        set out(width) "-width [lindex $args 6]"
+        set out(tags) [lindex $args 7]
+        switch $type {
+            "listbox" { set out(shape) twocorners }
+            "msg" { set out(shape) twospikes }
+            default { set out(shape) onecorner }
+        }
+        set out(capstyle) $defaultCapstyle
+    } elseif { "obj" eq $type } {
+        check_argc_exact 8 $argc
+        set out(pattern) "-dash \"[lindex $args 5]\""
+        set out(width) "-width [lindex $args 6]"
+        set out(tags) [lindex $args 7]
+        set out(shape) "rectangle"
+        set out(capstyle) $defaultCapstyle
+    } elseif { "commentbar" eq $type } {
+        check_argc_exact 6 $argc
+        set out(pattern) ""
+        set out(width) ""
+        set out(tags) [lindex $args 5]
+        set out(shape) "line"
+        set out(capstyle) ""
+    }
+}
+
+proc get_poly_coords {pPtr} {
+    upvar 1 $pPtr p
+    array_to_vars p
+    switch $shape {
+        "twocorners" {
+            return "$x1 $y1 [expr $x2 - $corner] $y1 $x2 [expr $y1 + $corner] $x2 [expr $y2 - $corner] [expr $x2 - $corner] $y2 $x1 $y2 $x1 $y1"
+        }
+        "onecorner" {
+            return "$x1 $y1 [expr $x2 - $corner] $y1 $x2 [expr $y1 + $corner] $x2 $y2 $x1 $y2 $x1 $y1"
+        }
+        "twospikes" {
+            return "$x1 $y1 [expr $x2 + $corner] $y1 $x2 [expr $y1 + $corner] $x2  [expr $y2 - $corner] [expr $x2 + $corner] $y2 $x1 $y2 $x1 $y1"
+        }
+        "rectangle" {
+            return "$x1 $y1 $x2 $y1 $x2 $y2 $x1 $y2 $x1 $y1"
+        }
+        "line" {
+            return "$x1 $y1 $x2 $y2"
+        }
+        default {
+            throw "get_poly_coords()" "[info level -2] >> unknown shape $shape"
+        }
+    }
+}
+
+proc bang_get_tags { obj outPtr } {
+    upvar 1 $outPtr out
+    #The C backend sends $obj with trailing 0x, however the legacy version
+    #would create tags with sprintf(...%lx..)(i.e.: no leading 0x). Here we
+    #remove the trailing 0x.
+    #TODO: better options would be to find a a better type from C instead
+    set obj [string replace $obj 0 1 "" ];
+    append out(object) $obj OBJ
+    append out(base) $obj BASE
+    append out(button) $obj BUT
+    append out(label) $obj LABEL
+}
+
+set iolets {outlet inlet iemgui_outlet iemgui_inlet}
+set poly_types {obj floatatom symbolatom listbox msg commentbar}
+set cnv_coords_types "$::poly_types $::iolets"
+
+proc ::pdtk_canvas::create {args} {
+    #puts "pdtk_canvas::create got: $args"
+    set docmds ""
+    check_argc_least 2 $args
```

Comment:
is there any specific reason to manually check the number of arguments rather than doing `proc ::pdtk_canvas::create {type args} { ... }`?

Reply:
not really